packages/flutter_tools/lib/src/web/bootstrap.dart

```diff
@@ -4,6 +4,320 @@
+          // https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState
+          if (this.readyState == 4 && this.status == 200 || this.status == 304) {
+            var scripts = JSON.parse(this.responseText);
+            var numToLoad = 0;
+            var numLoaded = 0;
+            for (var i = 0; i < scripts.length; i++) {
+              var script = scripts[i];
+              if (script.id == null) continue;
+              var src =
+                  window.\$dartAppNameToMetadata.get(appName).currentDirectory +
+                  script.src.toString();
+              var oldSrc = window.\$dartLoader.moduleIdToUrl.get(script.id);
+              // Only compare the search parameters which contain the cache
+              // busting portion of the uri. The path might be different if the
+              // script is loaded from a different application on the page.
+              if (new URL(oldSrc).search == new URL(src).search) continue;
+
+              // We might actually load from a different uri, delete the old one
+              // just to be sure.
+              window.\$dartLoader.urlToModuleId.delete(oldSrc);
+
+              window.\$dartLoader.moduleIdToUrl.set(script.id, src);
+              window.\$dartLoader.urlToModuleId.set(src, script.id);
+
+              if (window.\$dartJITModules) {
+              // Simply invalidate the import and the corresponding module will
+              // be lazily loaded.
+              dart_library.invalidateImport(script.id);
+              continue;
+              } else {
+                numToLoad++;
+              }
+
+              var el = document.getElementById(script.id);
+              if (el) el.remove();
+              el = window.\$dartCreateScript();
+              el.src = policy.createScriptURL(src);
+              el.async = false;
+              el.defer = true;
+              el.id = script.id;
+              el.onload = function() {
+                numLoaded++;
+                if (numToLoad == numLoaded) cb();
+              };
+              document.head.appendChild(el);
+            }
+            // Call `cb` right away if we found no updated scripts.
+            if (numToLoad == 0) cb();
+          }
+        };
+        xhttp.open("GET",
+          window.\$dartAppNameToMetadata.get(appName).fetchEtagsUri, true);
+        let sdk = dart_library.import("dart_sdk", appName);
+        let developer = sdk.developer;
+        if (developer._extensions.containsKey("ext.flutter.disassemble")) {
+          developer.invokeExtension("ext.flutter.disassemble", "{}").then(() => {
+            // TODO(b/204210914): we should really be clearing all statics for all
+            // apps, but for now we just do it for flutter apps which we recognize
+            // based on this extension.
+            sdk.dart.hotRestart();
+            xhttp.send();
+          });
+        } else {
+          xhttp.send();
+        }
+      });
+    }
+  }
+})();
```

Comment:
What does the `$ddcModuleLoaderUrl` load? Because it seems the entirety of DDC loading has just been appended here?

Is there any chance that the previous ~200 lines of code are extracted to a separate file, similar to "require.js" (call it "ddc.js" or whatever) and then the bootstrap logic stays kinda small?

From about:

```
(function() {
  let appName = "$entrypoint";
```

to this comment.

Reply:
This is unfortunately not the entirety of DDC's loading. We want to move most of this logic into `ddc_module_loader.js`, but that will be a slow and heavily-tested process. Some of the code here is responsible for hot restart internally, so I've stripped it for now (until I formally port it over). Hopefully it's cleaner now!

Reply:
Yes, the more code you can have in `ddc_module_loader.js` (vended from the SDK) and tested, the better! :P Ideally `ddc_module_loader.js` would be similar to `require.js`, in which it only requires as initialization parameters some of the things that we're substituting into JS here.

Reply:
Exactly! ddc_module_loader.js should encompass what is now parts of client.js in dwds and require.js. That is, handling hot restart, hot reload, lazy module loading, metrics collection, and debugger hooks.